src/Language/PureScript/TypeChecker.hs

```diff
@@ -434,31 +434,116 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
     return $ Module ss coms mn decls' (Just exps)
   where
+  qualify' :: a -> Qualified a
+  qualify' = Qualified (Just mn)
+
+  annotateSuperClasses 
+    :: (Qualified (ProperName 'ClassName) -> S.Set (Qualified (ProperName 'ClassName)))
+    -> M.Map (Qualified (ProperName 'ClassName)) DeclarationRef
+    -> MultipleErrors
+    -> m ()
+  annotateSuperClasses transitiveSuperClassesFor classMap (MultipleErrors errs) = 
+    throwError (MultipleErrors (map annotateError errs))
+    where
+      annotateError (ErrorMessage hints (TransitiveExportError ref@(TypeClassRef _ className) _)) =
+        let superClasses = transitiveSuperClassesFor (qualify' className)
+            unexportedClasses = S.difference superClasses moduleClassExports
+        in ErrorMessage hints 
+          $ TransitiveExportError ref 
+          $ mapMaybe (`M.lookup` classMap) 
+          $ toList unexportedClasses
+      annotateError err = err
+
+  moduleClassExports :: S.Set (Qualified (ProperName 'ClassName))
+  moduleClassExports = S.fromList $ mapMaybe (\x -> case x of
+     TypeClassRef _ name -> Just (qualify' name)
+     _ -> Nothing) exps
+
+  untilSame :: Eq a => (a -> a) -> a -> a
+  untilSame f a = let a' = f a in if a == a' then a else untilSame f a'
 
   checkMemberExport :: (Type -> [DeclarationRef]) -> DeclarationRef -> m ()
   checkMemberExport extract dr@(TypeRef _ name dctors) = do
     env <- getEnv
-    case M.lookup (Qualified (Just mn) name) (typeSynonyms env) of
-      Nothing -> return ()
-      Just (_, ty) -> checkExport dr extract ty
-    case dctors of
-      Nothing -> return ()
-      Just dctors' -> for_ dctors' $ \dctor ->
-        case M.lookup (Qualified (Just mn) dctor) (dataConstructors env) of
-          Nothing -> return ()
-          Just (_, _, ty, _) -> checkExport dr extract ty
-    return ()
+    for_ (M.lookup (qualify' name) (typeSynonyms env)) $ \(_, ty) ->
+      checkExport dr extract ty
+    for_ dctors $ \dctors' ->
+      for_ dctors' $ \dctor ->
+        for_ (M.lookup (qualify' dctor) (dataConstructors env)) $ \(_, _, ty, _) ->
+          checkExport dr extract ty
   checkMemberExport extract dr@(ValueRef _ name) = do
-    ty <- lookupVariable (Qualified (Just mn) name)
+    ty <- lookupVariable (qualify' name)
     checkExport dr extract ty
   checkMemberExport _ _ = return ()
 
+  checkSuperClassExport
+    :: (Qualified (ProperName 'ClassName) -> S.Set (Qualified (ProperName 'ClassName)))
+    -> M.Map (Qualified (ProperName 'ClassName)) DeclarationRef
+    -> DeclarationRef
+    -> m ()
+  checkSuperClassExport superClassesFor classMap dr@(TypeClassRef _ className) = do
+    let superClasses = superClassesFor (qualify' className)
+        unexportedClasses = S.difference superClasses moduleClassExports
+    unless (null unexportedClasses)
+      . throwError . errorMessage' (declRefSourceSpan dr)
+      . TransitiveExportError dr
+      $ mapMaybe (\n -> M.lookup n classMap) (toList unexportedClasses)
```

Comment:
It seems to me that there's a bit of a code smell here as we chuck away the second argument of `TransitiveExportError` here almost immediately, i.e. once we reach `annotateSuperClasses`. Have you considered consolidating `annotateSuperClasses` and this into a single function?

Reply:
Great catch. Merged.